token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs

```diff
@@ -176,3 +163,14 @@ computeCurrentCount TokenLimitConfig {..} lastServiced countThen now =
    in if willOverflow
         then tokenLimitConfigMaxTokens
         else min tokenLimitConfigMaxTokens totalCount
+
+tryModifyMVar :: MVar a -> (a -> IO (a, b)) -> IO (Maybe b)
+tryModifyMVar m io =
+  uninterruptibleMask $ \restore -> do
```

Comment:
I think https://www.fpcomplete.com/haskell/tutorial/exceptions/ is a great read about this still. If you would like to deal with Haskell concurrency for some long time, I suggest strongly investing in understanding this (and the above linked Control.Exception doc on interruptible).

TLDR in Haskell a thread running some code can be excepted at "any" point (well, if it doesn't malloc, and runs in a tight computation loop, then it can't unless you explicitly use `yield`... but that's rare). 

(What is an interrupt? More specifically it is an async Exception, which is just an exception thrown asynchronously, that is to an other thread (like with throwTo). The Async lib for example uses this under the hood of `cancel` and all its combinators like concurrently etc)

Now, if you have piece of code that needs to uphold some invariant (like with modifyMVar, I either leave the MVar as-is, or write back the modified content, but never accidentally leave the MVar empty, which would then likely deadlock subsequent users), then you need to think if/where interrupts can occur. Usually what you want, you mask interrupts (that is, disable receiving them) for the setup of your handling code, and restore interruptibility for the action (once the handlers are set up).

Now, 'mask' disables interrupts, _but_ they can still happen at specific points (the "interruptible" points IIRC), which usually happen around "blocking" operations (from the app perspective - the runtime itself doesn't block but yields control to other stuff). This is to prevent deadlocks and whatever.

'uninterruptibleMask' really masks all interrupts, but then it means if your masked computation gets in a loop or something, there's no way to kill it. For example, in server code you usually want to serve requests from users in an Async and put some hard upper bound on execution time (via race + threadDelay). But if the computation is uninterruptibly masked, then it can't receive the async exception that kills it, and you are deadlocked. So don't use that.

(unliftio, rio use uninterruptible mask in their versions of `finally` etc, to guarantee completion of resource cleanup, taking the side of rather deadlocking than leaking resources. I lean to this approach as well, but it has some tricky downsides in some situations... it is kind of all-in, don't mix different semantics, either stick to base or full-in on unliftio)